README
mps-youtube
===========

.. image:: http://badge.fury.io/py/mps-youtube.png
    :target: https://pypi.python.org/pypi/mps-youtube
.. image:: https://pypip.in/d/mps-youtube/badge.png
    :target: https://pypi.python.org/pypi/mps-youtube


Features
--------
- Search and play audio/video from YouTube
- Search and import YouTube playlists
- Create and save local playlists
- Download audio/video
- Works with Python 2.7+ and 3.x
- Works with Windows, Linux and Mac OS X 
- Requires mplayer or mpv

This project is based on `mps <https://github.com/np1/mps>`_, a terminal based program to search, stream and download music.  This implementation uses YouTube as a source of content and can play and download video as well as audio.  The `pafy <https://github.com/np1/pafy>`_ library handles interfacing with YouTube.

Screenshots
-----------

Search
~~~~~~

.. image:: http://np1.github.io/mpsyt-images/mpsyt-search.png

Local Playlists
~~~~~~~~~~~~~~~

.. image:: http://np1.github.io/mpsyt-images/mpsyt-playlists.png


Playback (video)
~~~~~~~~~~~~~~~~

.. image:: http://np1.github.io/mpsyt-images/mpsyt-vid-playback.png

Download
~~~~~~~~

.. image:: http://np1.github.io/mpsyt-images/mpsyt-download.png


Installation
------------

Using `pip <http://www.pip-installer.org>`_::
    
    [sudo] pip install mps-youtube

Mac OS X installation notes
~~~~~~~~~~~~~~~~~~~~~~~~~~~
    
Install mplayer with `MacPorts <http://www.macports.org>`_::

    sudo port install MPlayer


Windows installation notes
~~~~~~~~~~~~~~~~~~~~~~~~~~~

Install the python `colorama <https://pypi.python.org/pypi/colorama>`_ module to get colors (optional)::

    pip install colorama

Download mplayer for your CPU type from the "Build Selection table" `here
<http://oss.netfarm.it/mplayer-win32.php>`_. 

Extract the mplayer.exe file, saving it to your mpsyt folder


Upgrading
---------

It is recommended you update to the latest version.

Upgrade pip installation::

    [sudo] pip install mps-youtube --upgrade

Usage
-----

mps-youtube is run on the command line using the command::
    
    mpsyt
    
Enter ``h`` from within the program for help.

Searching
~~~~~~~~~

Enter your search term by prefixing it with a ``.`` or ``/`` character.
    e.g., ``/daft punk``

Prefix with ``..`` or ``//`` to search for YouTube playlists.  
    e.g., ``..linux tutorials``

Enter ``n`` or ``p`` to go to the next / previous page of results


When a list of search results or a saved playlist is shown, you can use the
following commands:

Information
~~~~~~~~~~~

``i 3`` to view info on item 3

Downloading
~~~~~~~~~~~

``d 3`` to download item 3 (will open new menu)
``da 3-5, 7`` to download audio streams of items 3, 4, 5 and 7
``dv 3-5, 7`` to download video streams of items 3, 4, 5 and 7

View Uploads / Related Videos
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

``u 3`` to view more videos from uploader of item 3

``r 3`` to view videos related to item 3

Playback
~~~~~~~~

``all`` to play all displayed items

``1,2,3`` to play items 1 2 and 3

``2-4,6,6-3`` to play items 2, 3, 4, 6, 6, 5, 4, 3

Note: The commands ``shuffle`` and ``repeat`` can be inserted at the start or
end of any of the above to enable those play modes: eg, ``shuffle 1-4`` or
``2-4,1 repeat`` 

Editing
~~~~~~~
``rm 1,5`` to remove items 1 and 5.

``rm 1,2,5-7`` to remove items 1,2 and 5-7.

``rm all`` to remove all items

``sw 1,3`` to swap the position of items 1 and 3

``mv 1,3`` to move items 1 to postion 3

Playlist commands
~~~~~~~~~~~~~~~~~

``add 1,2,3`` to add items 1,2 and 3 to the current playlist. 

``add 1-4,6,8-10`` to add items 1-4, 6, and 8-10 to the current playlist
    
``add 1-4,7 <playlist_name>`` to add items 1-4 and 7 to a saved playlist.  A
new playlist will be created if the given name doesn't already exist.

``vp`` to view the current playlist (then use rm, mv and sw to modify it)

``ls`` to list your saved playlists

``open <playlist_name or ID>`` to open a saved playlist as the current playlist 

``view <playlist_name or ID>`` to view a playlist (leaves current playlist intact)

``play <playlist_name or ID>`` to play a saved playlist directly.

``save`` or ``save <playlist_name>`` to save the currently displayed items as a
stored playlist on disk

``rmp <playlist_name or ID>`` to delete a playlist from disk

``mv <old_name or ID> <new_name>`` to rename a playlist

``q`` to quit

``h`` for help


Advanced Tips
-------------

Playlist Name Completion
~~~~~~~~~~~~~~~~~~~~~~~~

When using ``open``, ``view`` or ``play``  to access a playlist, you can enter
the first few characters instead of the whole name.  The first alphabetically
matching playlist will be opened / displayed.

Invocation
~~~~~~~~~~

To play a saved playlist when invoking mps-youtube use the following command:

``mpsyt play <playlistname>``

This also works for other commands, eg:

``mpsyt .mozart`` to search .

``mpsyt view <playlistname>`` to view a saved playlist.

``mpsyt ls`` to list saved playlists.

``mpsyt open moz`` to open a saved playlist called mozart.

``mpsyt playurl <url>`` to play a YouTube video by url.

``mpsyt dlurl <url>`` to download a YouTube video by url.

Note: If the url contains & symbols you may need to quote it with ""

Specifying Ranges
~~~~~~~~~~~~~~~~~

When selecting items for playback, removing or adding you can use ``5-`` to 
select items 5 upward and ``-5`` to select up to item 5.  This can be included
with other choices so for example: ``5,3,7-,-2``.  You can also use spaces
instead of commas eg. ``5 3 7- -2``.

Changing Player Application
~~~~~~~~~~~~~~~~~~~~~~~~~~~

To set a different player, from within mps-youtube::

    set player mpv

or::

    set player mplayer

Other Configuration
~~~~~~~~~~~~~~~~~~~

To view configuration, enter ``set`` and to change any item enter: 
``set <item> <value>``.  This can be used to change the download path (DDIR)
and will persist after exiting the program.  To reset all settings to default,
use ``set all default`` or for a single item, ``set <item> default``

Search All Categories
~~~~~~~~~~~~~~~~~~~~~

To search all YouTube categories (instead of just music), enter:: 
    
    set search_music false

List YouTube User Uploads / Playlists 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To list the uploaded videos of a YouTube user:

    user <username>

To search videos uploaded by a YouTube user:

    user <username>/<search terms>

To list playlists created by a YouTube user:

    userpl <username> 

Access YouTube Video by URL
~~~~~~~~~~~~~~~~~~~~~~~~~~~

    url <YouTube video url or id>

Open YouTube Playlist by URL
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

    pl <YouTube playlist url or id>

Show Entire YouTube Playlist Contents
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When a YouTube playlist is displayed:

    dump

Show Video Content / Fullscreen Mode
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To view video instead of audio, enter::

    set show_video true

To play video content in fullscreen mode::

    set fullscreen true

Note: The above two settings can be overridden on a per-case basis by adding
``-w``, ``-f`` or ``-a`` to your selection to play the items in windowed, fullscreen
or audio-only modes.  E.g., ``1,4,7 -f``


CHANGELOG
<!--next-version-placeholder-->

Version 0.2.8
17 February 2018

Features:
---------

- Lastfm scrobbling (@rien333) (#758)
- `splaylist` and `suser` commands for Spotify (@ritiek) (#678)
- Autoplay support (@nikhilweee) (#594)
- `genpl` command (@nishanthkarthik) (#642)
- `video_format` configuration setting (@Vrihub) (#656)
- `live` command for live streams (@Razesdark) (#635)
- Add time column (@akaWolf) (#582)
- Add video ID column (@Razesdark) (#627)
- Allow filtering search by duration (@jas32096) (#548)
- Allow filtering search by date (@jas32096) (#553)
- Add `history recent` to display history without duplicates (@srvanrell) (#510)
- `--no-textart` to disable ascii art (for screenreaders) (@Gongreg) (#517)
- `set_title` option to disable setting window title (@ardrabczyk) (#498)
- Add `s` command for getting stream information (@ritiek) (#675)
- Add option to disable mpris (@hrnr) (#728)
- Generate playlist from video description (@Razesdark) (#649)

Enhancements:
-------------
- Store playlists as m3u files instead of binary (@kraetzin) (#682)
- Show replies when displaying comments (@paulfertser) (#677)
- Use https urls for album art (@vszakats) (#574)
- When using "repeat" with mpv, keep player running (@ritiek) (#660)
- Display artist/album in mpris using lastfm (@vn-ki) (#739)
- Use channel.list API for user search (@Razesdark) (#628)
- Maintain volume across songs with mpv (@Razesdark) (#720)
- Prevent mpris from crashing the main process (@hrnr) (#728)

Bug fixes:
----------
- Fix mpris seeking with recent mpv vesions (@mg6) (#687)
- Fix removing video from multi-page playlist (@kraetzin) (#695)
- Skip unavailable tracks (@ritiek) (#734)
- Correctly display minutes/seconds (@Laxa) (#748)
- Handle `null` time-pos from mpv (@pritambaral) (#519)
- Fix da/dv command regex (@Vrihub) (#691)
- Skip channels in video/playlist searches (@ritiek) (#664)
- Fix divide by zero error (@ids1024) (#620)

Use `git log v0.2.7.1..v0.2.8` to see all changes, including various minor
fixes, documentation changes, etc. not included here.

-------------------------------------------------------------------------------

Version 0.2.7.1
6 July 2016

Bug fixes:
----------
- Fix pickle error (@ids1024) (#503)
- Install LICENSE, README.md, and CHANGELOG as package_data (@ids1024)
- Update youtube-dl in py2exe build (@ids1024)

-------------------------------------------------------------------------------

Version 0.2.7
27 June 2016

Features:
---------
- Setting for default audio format (@nishanthkarthik) (#71 #466 #467)
- Search history with `history` command (@kraetzin) (#344 #472)
- Add syntax for repeating a track several times (@ghallak) (#437 #490)
- Reverse command (@kraetzin) (#422 #473)
- New `daurl <url>` command (@maricn) (#402)

Enhancements:
-------------
- Pass `--no-ytdl` to mpv (@ids1024) (#492)
- Do not suppress mpv output in debug mode (@ids1024)
- Print traceback in debug mode (@ids1024)
- Speed up `checkupdate` by avoiding redirects (@ids1024) (#454)
- Pass `--play-and-exit` if vlc is used as a player (@ids1024)`

Bug fixes:
----------
- Fix error with python 3.3 (@ids1024) (#414)
- Fix maxres bug (@ids1024) (#430)
- Fix playlist cache bug in info command (@ids1024) (#468)
- Fix usersearch argument error (@ids1024) (#469 #482)
- Fix progress bar with new mpv versions (@ids1024) (#470 #485)
- Fix syntax error with set command (@ids1024) (#419)
- Fix syntax error with -w, -f, -a (@ids1024) (#420)
- Fix clipboard support (@ids1024) (#461)
- Fix mpv version checking on windows (@ids1024) (#397)
- Add youtube_dl as dependency to Dockerfile (@PI-Victor) (#481)

Internal api changes:
---------------------
- The code is now split between several files (@ids1024) (#208)
- Add initial content api, for now only used for comments (@ids1024)
- Begin documenting internal apis using sphinx (@ids1024)

-------------------------------------------------------------------------------

Version 0.2.6
1 January 2016

Features:
- Add browserplay command to play video in browser (rjvani) (#355 #390)
- Make url command support multiple url, and allow loading urls from file
    (paulnicolet) (#90 #391)
- Make youtube id available to download command (ids1024) (#375)
- Make pagination work properly for all lists of songs (local playlists, etc.)
    and make them all support the dump command (ids1024) (#201)
- Show warning when adding duplicate track to playlist (BensonQiu) (#374 #383)

Bug fixes:
- Support playlists longer than 200 items (ids1024) (#192)
- Make dapl download complete playlist (JKatzwinkel) (#294 #325)
- Use pyperclip instead of xerox, fixing clipboard on linux (ids1024) (#311)
- Fix error message on windows when video title has & (ids1024) (#321)
- Run notifier command in background (hrnr) (#339)
- Correctly handle -v option of play (TimoDritschler) (#352)
- Make it possible to use a pager with --help (livingBEEF) (#349)
- Allow mplayer for https streams if version is new enough
    (ids1024 Evidlo paddatrapper) (#315 #336 #317)
- Use mpv.com instead of mpv.exe on windows (ids1024)

Internal API changes:
- Split code into several files (still a work in progress) (ids1024) (#208)
- Use a python decorator to declare commands (ids1024)
- Create a paginatesongs() function for handling the pagination and dump
  command uniformly as mentioned in features (#399) (ids1024)
- Various others

There have been various other small bug fixes and internal API changes. Use
`git log v0.2.5..0.2.6` to see all of them.

-------------------------------------------------------------------------------

Version 0.2.5
1 June 2015

- Support python 3 only (will not run with python 2)
- Fix progress bar bug on Windows (#215, #223)
- Fix bug in playlist search (#282)
- Add user_order setting for separate order of user searches (lol768) (#277)
- Add mix command (lol768) (#241, #283)
- Add page count with total number of pages and current page (JKatzwinkel)
- Fix UnicodeEncodeError on non-unicode terminals

setup_py2exe.py
#!/usr/bin/python

""" setup.py for mps-youtube.

https://np1.github.com/mps-youtube

python setup.py sdist bdist_wheel
"""

try:
    from setuptools import setup

except ImportError:
    from distutils.core import setup

import py2exe

setup(
    name="mps-youtube",
    version="0.2.1",
    description="Terminal based YouTube player and downloader",
    keywords=["video", "music", "audio", "youtube", "stream", "download"],
    author="np1",
    author_email="np1nagev@gmail.com",
    url="http://github.com/np1/mps-youtube",
    download_url="https://github.com/np1/mps-youtube/tarball/master",
    packages=['mps_youtube'],
    entry_points=dict(console_scripts=['mpsyt = mps_youtube:main.main']),
    console=['mpsyt'],
    install_requires=['Pafy >= 0.3.72'],
    package_data={"": ["LICENSE", "README.rst", "CHANGELOG"]},
    classifiers=[
        "Topic :: Utilities",
        "Topic :: Internet :: WWW/HTTP",
        "Topic :: Multimedia :: Sound/Audio :: Players",
        "Topic :: Multimedia :: Video",
        "Environment :: Console",
        "Environment :: Win32 (MS Windows)",
        "Environment :: MacOS X",
        "Operating System :: POSIX :: Linux",
        "Operating System :: MacOS",
        "Operating System :: MacOS :: MacOS 9",
        "Operating System :: MacOS :: MacOS X",
        "Operating System :: Microsoft",
        "Operating System :: Microsoft :: Windows :: Windows 7",
        "Operating System :: Microsoft :: Windows :: Windows XP",
        "Operating System :: Microsoft :: Windows :: Windows Vista",
        "Intended Audience :: End Users/Desktop",
        "Programming Language :: Python",
        "Programming Language :: Python :: 2",
        "Programming Language :: Python :: 3",
        "Programming Language :: Python :: 2.7",
        "Programming Language :: Python :: 3.3",
        "Programming Language :: Python :: 3.4",
        "Development Status :: 5 - Production/Stable",
        "License :: OSI Approved :: GNU General Public License v3 (GPLv3)"
    ],
    long_description=open("README.rst").read(),
    options={"py2exe": {"excludes": "readline, win32api, win32con, xerox",
	                "bundle_files": 1}},
    zipfile=None
)
